### ISSUE TITLE
Exported-change-not-reimported error after deleting the last alias

### ISSUE THREAD

**Issue by khandelwalpiyush** [lithnet/googleapps-managementagent#8]
Not sure if this is a MA bug. 

Scenario

1) Make the aliase of a group primary mail address and then delete the old primary mail address in same operation
2) MIM will try to export a rename (due to our rules) and delete the alias
3) Export completes successfully
4) Checking google GUI the alias is also deleted and primary mail address has changed

But on Delta Import we get "exported-change-not-reimported" as it thinks we still need to delete mail alias

delta XML

```xml
<?xml version="1.0" encoding="utf-8"?>
<cs-entry-changes>
  <object-change>
    <modification-type>Update</modification-type>
    <dn>mailprefix4-l@d2-monash-edu.ga-staff-dev.monash.edu</dn>
    <object-class>group</object-class>
    <anchor-attributes>
      <anchor-attribute>
        <name>id</name>
        <value>03o7alnk0xktbdz</value>
      </anchor-attribute>
    </anchor-attributes>
    <attribute-changes>
      <attribute-change>
        <name>description</name>
        <modification-type>Replace</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Add</modification-type>
            <value>distgroup</value>
          </value-change>
        </value-changes>
      </attribute-change>
      <attribute-change>
        <name>primaryEmail</name>
        <modification-type>Replace</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Add</modification-type>
            <value>mailprefix4-l@d2-monash-edu.ga-staff-dev.monash.edu</value>
          </value-change>
        </value-changes>
      </attribute-change>
      <attribute-change>
        <name>name</name>
        <modification-type>Replace</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Add</modification-type>
            <value>distgroup</value>
          </value-change>
        </value-changes>
      </attribute-change>
      <attribute-change>
        <name>aliases</name>
        <modification-type>Update</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Delete</modification-type>
            <value>mailprefix41-l@d2-monash-edu.ga-staff-dev.monash.edu</value>
          </value-change>
        </value-changes>
      </attribute-change>
    </attribute-changes>
  </object-change>
</cs-entry-changes>
```
On subsequent export it fails again as it is trying to delete an already deleted alias. Only Full Import Delta Sync clears the error as it confirms the delete

**Comment by ryannewington** [comment]
@khandelwalpiyush 
That delta file does have the alias delete in it.

```xml
<attribute-change>
        <name>aliases</name>
        <modification-type>Update</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Delete</modification-type>
            <value>mailprefix41-l@d2-monash-edu.ga-staff-dev.monash.edu</value>
          </value-change>
        </value-changes>
      </attribute-change>
```

Can you confirm that aliases was the correct attribute?

**Comment by khandelwalpiyush** [comment]
aliases is the correct attribute.. Add has no issues and confirms on DI

```xml
      <attribute-change>
        <name>aliases</name>
        <modification-type>Update</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Add</modification-type>
            <value>mailprefix5-l@d2-monash-edu.ga-staff-dev.monash.edu</value>
          </value-change>
        </value-changes>
      </attribute-change>
```
On doing just a alias delete (no other operation) it also works


```xml
<?xml version="1.0" encoding="utf-8"?>
<cs-entry-changes>
  <object-change>
    <modification-type>Update</modification-type>
    <dn>mailprefix51-l@d2-monash-edu.ga-staff-dev.monash.edu</dn>
    <object-class>group</object-class>
    <anchor-attributes>
      <anchor-attribute>
        <name>id</name>
        <value>00z337ya3wiun09</value>
      </anchor-attribute>
    </anchor-attributes>
    <attribute-changes>
      <attribute-change>
        <name>aliases</name>
        <modification-type>Update</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Delete</modification-type>
            <value>mailprefix5-l@d2-monash-edu.ga-staff-dev.monash.edu</value>
          </value-change>
        </value-changes>
      </attribute-change>
    </attribute-changes>
  </object-change>
</cs-entry-changes>
```

But the combined operation of renaming the primary and then deleting the primary which has become the alias (not yet exported to google) is not confirmed (the XML in the original post) .. although the operation is successful on google end

**Comment by khandelwalpiyush** [comment]
How to reproduce

- Group with PK: primary@d2-monash-edu.ga-staff-dev.monash.edu
- Alias: newprimary@d2-monash-edu.ga-staff-dev.monash.edu

Now we switch alias to PK and delete the PK under same operation. Thus

- New PK (Rename operation): newprimary@d2-monash-edu.ga-staff-dev.monash.edu
- Delete of alias: newprimary@d2-monash-edu.ga-staff-dev.monash.edu

I think if we look at above.. it is wrong as the delete operation of alias won't work correctly as its becoming the PK and thus nothing to delete.. 

![groups-email-rename](https://cloud.githubusercontent.com/assets/19180629/18236334/4cacc758-7368-11e6-8ea1-a68c2793aba2.JPG)

On import it is expecting a null for alias but the delta import file contains the email address. 

![groups-email-import](https://cloud.githubusercontent.com/assets/19180629/18236345/651f6890-7368-11e6-8546-2ca5da94503b.JPG)

Might just be the way I am doing the operation. If not considered a bug please close it.

**Comment by ryannewington** [comment]
Thanks, that helps a lot. Can you provide the delta file from the operation you logged above?

**Comment by khandelwalpiyush** [comment]
```xml
<?xml version="1.0" encoding="utf-8"?>
<cs-entry-changes>
  <object-change>
    <modification-type>Update</modification-type>
    <dn>newprimary@d2-monash-edu.ga-staff-dev.monash.edu</dn>
    <object-class>group</object-class>
    <anchor-attributes>
      <anchor-attribute>
        <name>id</name>
        <value>01baon6m2xe458k</value>
      </anchor-attribute>
    </anchor-attributes>
    <attribute-changes>
      <attribute-change>
        <name>description</name>
        <modification-type>Replace</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Add</modification-type>
            <value>distgroup</value>
          </value-change>
        </value-changes>
      </attribute-change>
      <attribute-change>
        <name>primaryEmail</name>
        <modification-type>Replace</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Add</modification-type>
            <value>newprimary@d2-monash-edu.ga-staff-dev.monash.edu</value>
          </value-change>
        </value-changes>
      </attribute-change>
      <attribute-change>
        <name>name</name>
        <modification-type>Replace</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Add</modification-type>
            <value>distgroup</value>
          </value-change>
        </value-changes>
      </attribute-change>
      <attribute-change>
        <name>aliases</name>
        <modification-type>Update</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Delete</modification-type>
            <value>primary@d2-monash-edu.ga-staff-dev.monash.edu</value>
          </value-change>
        </value-changes>
      </attribute-change>
    </attribute-changes>
  </object-change>
</cs-entry-changes>
```

**Comment by ryannewington** [comment]
I suspect this might be a FIM bug. To confirm, can you repeat this process with ANOTHER alias on the object that does not change throughout these operations. So 2 aliases on the object, and one of them becomes the primary.

**Comment by khandelwalpiyush** [comment]
On the latest build the following was executed successfully

1) PK: aliase@d2-monash-edu.ga-staff-dev.monash.edu
2) Aliase: mailprefix5-l@d2-monash-edu.ga-staff-dev.monash.edu ; aliase2@d2-monash-edu.ga-staff-dev.monash.edu

Then one of the aliase was set PK and the original PK was deleted.. End result.. 

1) PK: mailprefix5-l@d2-monash-edu.ga-staff-dev.monash.edu
2) Aliase: alias2@d2-monash-edu.ga-staff-dev.monash.edu

**Comment by ryannewington** [comment]
What about the original scenario? Does that work with the build I gave you?

**Comment by khandelwalpiyush** [comment]
Yes.. Works with original scenario

Changed PK from mailprefix5-l@d2-monash-edu.ga-staff-dev.monash.edu to aliases@d2-monash-edu.ga-staff-dev.monash.edu and deleted mailprefix5-l@d2-monash-edu.ga-staff-dev.monash.edu  from aliases in one operation

DIDS confirmed delete of aliase

```xml

<?xml version="1.0" encoding="utf-8"?>
<cs-entry-changes>
  <object-change>
    <modification-type>Update</modification-type>
    <dn>aliases@d2-monash-edu.ga-staff-dev.monash.edu</dn>
    <object-class>group</object-class>
    <anchor-attributes>
      <anchor-attribute>
        <name>id</name>
        <value>01ksv4uv4eofweb</value>
      </anchor-attribute>
    </anchor-attributes>
    <attribute-changes>
      <attribute-change>
        <name>description</name>
        <modification-type>Replace</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Add</modification-type>
            <value>distgroup</value>
          </value-change>
        </value-changes>
      </attribute-change>
      <attribute-change>
        <name>primaryEmail</name>
        <modification-type>Replace</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Add</modification-type>
            <value>aliases@d2-monash-edu.ga-staff-dev.monash.edu</value>
          </value-change>
        </value-changes>
      </attribute-change>
      <attribute-change>
        <name>name</name>
        <modification-type>Replace</modification-type>
        <data-type>String</data-type>
        <value-changes>
          <value-change>
            <modification-type>Add</modification-type>
            <value>distgroup</value>
          </value-change>
        </value-changes>
      </attribute-change>
      <attribute-change>
        <name>aliases</name>
        <modification-type>Delete</modification-type>
        <data-type>String</data-type>
        <value-changes />
      </attribute-change>
    </attribute-changes>
  </object-change>
</cs-entry-changes>
```

**Comment by ryannewington** [comment]
Ok. I'll apply the same change to user aliases as well. Can you just confirm that we don't have the same issue with members? (eg deleting the last member of a group causes export-change-not-reimported)

**Comment by khandelwalpiyush** [comment]
Yes had tested that before.. worked

```xml
<?xml version="1.0" encoding="utf-8"?>
<cs-entry-changes>
  <object-change>
    <modification-type>Update</modification-type>
    <dn>mailprefix-l@d2-monash-edu.ga-staff-dev.monash.edu</dn>
    <object-class>group</object-class>
    <anchor-attributes>
      <anchor-attribute>
        <name>id</name>
        <value>01rvwp1q1bqvuim</value>
      </anchor-attribute>
    </anchor-attributes>
    <attribute-changes>
      <attribute-change>
        <name>member</name>
        <modification-type>Update</modification-type>
        <data-type>Reference</data-type>
        <value-changes>
          <value-change>
            <modification-type>Delete</modification-type>
            <value>c-davinc@d2-monash-edu.ga-staff-dev.monash.edu</value>
          </value-change>
        </value-changes>
      </attribute-change>
    </attribute-changes>
  </object-change>
</cs-entry-changes>
```

**Comment by ryannewington** [comment]
Fixed in [v1.1.6093](https://github.com/lithnet/googleapps-managementagent/releases/tag/v1.1.6093)
